Expose reaction drawing and additional FPs in MinimalLib

ptosco opened the pull request:
- exposed reaction drawing in MinimalLib
- fixed a typo in the error message "JSON doesn't contain 'atoms' field, or it is not an array"
- replaced RapidJson `HasMember()` with `FindMember()` to avoid a duplicate lookup in case the member exists and can be accessed
- some cosmetic style changes (avoid multiple variable declarations on a single line, use curly bracket also for one-liner if clauses, use auto where possible)
- capitalized "greg Landrum" to "Greg Landrum" (well deserved)
- exposed other FPs in addition to Morgan and Pattern FPs in MinimalLib
- added relevant tests

greglandrum commented on Code/MinimalLib/jswrapper.cpp:
```diff
@@ -42,10 +47,9 @@ std::string draw_to_canvas_with_offset(JSMol &self, emscripten::val canvas,
     height = canvas["height"].as<int>();
   }
-  MolDraw2DJS *d2d = new MolDraw2DJS(width, height, ctx);
+  std::unique_ptr<MolDraw2DJS> d2d(new MolDraw2DJS(width, height, ctx));
```
nice

greglandrum reviewed (changes requested):
A couple of thoughts on this:
- This is the first step down the rabbit hole of supporting reactions in the MinimalLib. That's not necessarily a bad thing, but "minimal" isn't going to apply for much longer. Is there some way that we can make this functionality optional for people who are just working with molecules and want a smaller library?
- I think the new fingerprints would be easier to use if the additional fingeprint parameters were provided using JSON instead of overloads. So instead of having four overloads with increasing numbers of parameters there'd just be:
```
      .function("get_rdk_fp_as_uint8array",
                select_overload<emscripten::val(const JSMol &)>(
                    get_rdk_fp_as_uint8array))
      .function("get_rdk_fp_as_uint8array",
                select_overload<emscripten::val(const JSMol &, std::string)>(
                    get_rdk_fp_as_uint8array))
```
This would make the interface cleaner and would make code which uses it a lot easier to read. The two current fingerprinting functions could also have this JSON form exposed and then we could deprecate the other overloads.

ptosco commented:
@greglandrum I agree that Minimal is not really such anymore, and it will only get worse. We could add `cmake` `RDK_MINIMAL_FINGERPRINTS`, `RDK_MINIMAL_REACTIONS`, `RDK_MINIMAL_MCS` parameters defaulting to `OFF` and use precompiler directives for conditional compilation.
Agreed?

greglandrum commented:
> @greglandrum I agree that Minimal is not really such anymore, and it will only get worse. We could add `cmake` `RDK_MINIMAL_FINGERPRINTS`, `RDK_MINIMAL_REACTIONS`, `RDK_MINIMAL_MCS` parameters defaulting to `OFF` and use precompiler directives for conditional compilation. Agreed?

I think something like this might be a solution, but it would be good to take a look at a draft CMakeLists.txt first to see how tricky it's going to be to use.

I'd suggest we do that in a separate PR, after this one is merged.

ptosco commented:
OK for a follow-up PR to opt feature in and out MinimalLib.
I am all in favor of passing parameters as JSON. I’ll update this PR accordingly and deprecate the existing Morgan FP functions.

MichelML commented:
https://github.com/rdkit/rdkit/pull/5277#issuecomment-1132474536

If we go this route, happy to release various flavors of the minimal lib based on most common use cases as readily installable npm packages. Examples (if that makes any sense):

- molecule rendering only wasm module
- molecule rendering, identifiers, substruct lib, and descriptors calculation
- molecule rendering, descriptors, identifiers, descriptors, substruct lib, and reactions.

ptosco commented:
@MichelML Thanks for offering to do that. When I will add support for FMCS I will figure out how much the various components add to the overall size of the library in order to figure which components it make sense to opt in/out.

@greglandrum I have addressed your remarks and also consolidated the fingerprint and reaction drawing support across CFFI and JS. I also fixed a segfault occurring in CFFI when passing `nullptr` as `mol_pkl` or 0-sized `mol_pkl` to `get_*_fp()`.
This should now be good to go.

greglandrum reviewed (approved):
LGTM
